File: src/main/java/com/weeth/global/common/exception/BaseException.java
package com.weeth.global.common.exception;

import lombok.Getter;

@Getter
public abstract class BaseException extends RuntimeException {

    private final int statusCode;
    private final ErrorCodeInterface errorCode;

    public BaseException(int code, String message) {
        super(message);
        this.statusCode = code;
        this.errorCode = null;
    }

    public BaseException(int code, String message, Throwable cause) {
        super(message, cause);
        this.statusCode = code;
        this.errorCode = null;
    }

    public BaseException(ErrorCodeInterface errorCode, Throwable cause) {
        super(errorCode.getMessage(), cause);
        this.statusCode = errorCode.getStatus().value();
        this.errorCode = errorCode;
    }

    public BaseException(ErrorCodeInterface errorCode) {
        this(errorCode, null);
    }
}

File: src/main/kotlin/com/weeth/global/common/exception/BaseException.kt
package com.weeth.global.common.exception

abstract class BaseException(
    val errorCode: ErrorCodeInterface,
    message: String? = null,
    val data: Any? = null,
) : RuntimeException(message ?: errorCode.message) {
    val statusCode: Int get() = errorCode.status.value()
}
